### PR TITLE
Bugfix/whitlock/933 distributed closest point example crash

### LINKED ISSUE #933: Distributed closest-point example crashes when a rank has just 1 object point
The crash occurs when the code tries to access element 1 of a 1-element array.
The problem is likely in special case of a `BVH` for one point.  The code fakes a 2-point array.  This is fine internally, but external code, such as the example, may be unaware of it, leading to some inconsistencies.

To reproduce,
`
bin/data_collection_util -o box_2d_1 --min -1.5 -1.5 --max 1.5 1.5 --res 100 100 -p1
examples/quest_distributed_distance_query_ex --num-samples 1 -m box_2d_1.root
`

### PR DESCRIPTION
**Pull request by BradWhitlock**
# Summary

- This PR is a bugfix
- It Fixes #933 

The bug fix changes how the invalid bounding boxes are added in the BVH::initialize method so they are actually invalid (as in BoundingBox.clear()). Additional changes were made to ensure that the invalid BoundingBoxes remain invalid all the way into the bvh_traverse routine where we now check whether they are valid before calling user-supplied lambdas that determine whether a point is inside the bounding box. This results in bvh_traverse not calling the action lambda with an invalid current_node index, which is what had been happening to cause the crash indexing DistributedClosestPoint's pointsView out of bounds in the checkMinDist lambda. Finally, this bugfix extends the spin_bvh_test test case so it calls the BVH similar to how DistributedClosestPoints was calling it. The new test caused the crash on the unpatched version of Axom, while the new version does not exhibit the crash.

### CODE REVIEW

File: src/axom/spin/tests/spin_bvh.cpp [hunk 1]
```diff
@@ -1306,6 +1306,150 @@ void check_find_points_zip2d()
...
+ *        value to the checkMinDist lambda.
+ *
+ * \param [in] bvh The BVH object to use.
+ * \param [in] query_pts The array of query points to use against the BVH. 
+ */
+template <typename BVHType, typename PointType>
+void bvh_one_bbox_2d(BVHType& bvh,
+                     axom::Array<PointType>& points,
+                     axom::Array<PointType>& query_pts,
+                     bool none)
```
Comment (kennyweiss) [inline src/axom/spin/tests/spin_bvh.cpp:1355]: Please document the `none` and `points` parameters.

Based on the code below, it looks like the `none` parameter allows the code to either use the point or not.
Since the function takes an array, maybe it should be renamed to something like `bvh_compute_point_distances_2d` (or something better)?

File: src/axom/spin/tests/spin_bvh.cpp [hunk 2]
```diff
@@ -1306,6 +1306,150 @@ void check_find_points_zip2d()
...
+    double minSqDist {numerics::floating_point_limits<double>::max()};
+    /// Index within mesh of closest element
+    int minElem {-1};
+    /// MPI rank of closest element
+    int minRank {-1};
+  };
+
+  // Initialize the BVH with the points.
+  axom::IndexType npts = none ? 0 : points.size();
+  //axom::Array<BoxType> bboxes;
```
Comment (kennyweiss) [inline src/axom/spin/tests/spin_bvh.cpp:1375]: Remove commented out line?

File: src/axom/spin/tests/spin_bvh.cpp [hunk 3]
```diff
@@ -1306,6 +1306,150 @@ void check_find_points_zip2d()
...
+
+      // Traverse the tree, searching for the point with minimum distance.
+      it.traverse_tree(qpt, checkMinDist, traversePredicate);
+
+      // Save the index of the minElem.
+      results_view[idx] = curr_min.minElem;
+    });
+
+  // Make sure all of the indices we found are the expected value.
+  int expected_idx = none ? -1 : 0;
```
Comment (kennyweiss) [inline src/axom/spin/tests/spin_bvh.cpp:1426]: Should we expose `-1` as a named value in `BVH` rather than as a magic number?

**Review by kennyweiss — APPROVED** [review]
Thanks @BradWhitlock!

Looks like there were a bunch of unhandled corner cases for empty/uninitialized bounding boxes. Should we add an issue to take a pass through `BoundingBox` to document restrictions/preconditions for its functions and to add additional unit tests?

Please also ensure that we have sufficient unit test coverage for the new/modified functions in `primal::BoundingBox`

File: src/axom/spin/tests/spin_bvh.cpp [hunk 4]
```diff
@@ -1372,11 +1373,11 @@ void bvh_one_bbox_2d(BVHType& bvh,
 
   // Initialize the BVH with the points.
   axom::IndexType npts = none ? 0 : points.size();
-  //axom::Array<BoxType> bboxes;
-  axom::Array<BoxType> bboxes;
+  BoxType* bboxes = axom::allocate<BoxType>(npts + 1,
+                                            axom::execution_space<ExecSpace>::allocatorID());
```
Comment (kennyweiss) [inline src/axom/spin/tests/spin_bvh.cpp]: The following should also work:
```cpp
auto allocID = axom::execution_space<ExecSpace>::allocatorID();
axom::Array<BoxType> bboxes(npts, npts, allocID);

for(axom::IndexType i = 0; i < npts; i++)
{
    bboxes[i] = BoxType(points[i]);
}
```

Comment (kennyweiss) [reply src/axom/spin/tests/spin_bvh.cpp]: i.e. you can construct and allocate an array of the desired size using a desired allocator.

File: src/axom/spin/internal/linear_bvh/bvh_traverse.hpp [hunk 5]
```diff
@@ -93,8 +93,8 @@ AXOM_HOST_DEVICE inline void bvh_traverse(
     {
       BBoxType left_bin = inner_nodes[current_node + 0];
       BBoxType right_bin = inner_nodes[current_node + 1];
-      const bool in_left = B(p, left_bin);
-      const bool in_right = B(p, right_bin);
+      const bool in_left = left_bin.isValid() ? B(p, left_bin) : false;
+      const bool in_right = right_bin.isValid() ? B(p, right_bin) : false;
```
[the remark this answers is not in the document]
Comment (BradWhitlock) [reply src/axom/spin/internal/linear_bvh/bvh_traverse.hpp:97]: There probably is something of a performance hit and we probably only need to check it in the 0 or 1 box cases. I'll see about a change there.

File: src/axom/primal/geometry/BoundingBox.hpp [hunk 6]
```diff
@@ -485,8 +485,25 @@ template <typename OtherT>
...
+  else
+  {
+    PointType m0, m1;
+    for(int i = 0; i < NDIMS; i++)
+    {
+      m0[i] = static_cast<T>(bbox.getMin()[i]);
+      m1[i] = static_cast<T>(bbox.getMax()[i]);
+    }
+    this->setMin(m0);
+    this->setMax(m1);
```
[the remark this answers is not in the document]
Comment (BradWhitlock) [reply src/axom/primal/geometry/BoundingBox.hpp:511]: The addition of the OtherT template argument is what made this code necessary, even though it's kind of dumb. I initially tried code like m_min = bbox.m_min but that causes type conversion errors for some instantiations. The "*this = bbox" suggestion does too:

> /usr/WS2/whitlocb/Axom/axom/src/axom/core/../../axom/primal/geometry/BoundingBox.hpp(514): error: no operator "=" matches these operands
>             operand types are: axom::primal::BoundingBox<float, 3> = const axom::primal::BoundingBox<double, 3>

For the else case, I do not think we need to check isValid. In the else, "this" is invalid, which I viewed as a BoundingBox equivalent of zero/nothing. Taking the input bbox as the new value either results in another invalid bbox (no-op) or a good one if the input was valid.

File: src/axom/spin/policy/LinearBVH.hpp [hunk 7]
```diff
@@ -58,7 +58,10 @@ class LinearBVHTraverser
   {
     auto traversePref = [](const BoxType& l, const BoxType& r, const PointType& p) {
       double sqDistL = primal::squared_distance(p, l.getCentroid());
-      double sqDistR = primal::squared_distance(p, r.getCentroid());
+      // If the right bbox is not valid, return max. Otherwise, the invalid right
+      // bbox might actually win when we should ignore it.
+      double sqDistR = r.isValid() ? primal::squared_distance(p, r.getCentroid())
```
[the remark this answers is not in the document]
Comment (BradWhitlock) [reply src/axom/spin/policy/LinearBVH.hpp:63]: I think the centroid returned 0, which seemed ok to me. It looks like the Doxygen code would not indicate a isValid restriction.

```
/*!
   * \brief Returns the centroid (midpoint) of the bounding box.
   * \return Point at the bounding box centroid.
   */
```

File: src/axom/spin/tests/spin_bvh.cpp [hunk 8]
```diff
@@ -1306,6 +1306,150 @@ void check_find_points_zip2d()
...
+      static_cast<FloatType>(j) / static_cast<FloatType>(dims[1] - 1);
+    FloatType y = (1. - tj) * y0 + tj * y1;
+    for(int i = 0; i <= dims[0]; i++)
+    {
+      FloatType ti =
+        static_cast<FloatType>(i) / static_cast<FloatType>(dims[0] - 1);
+      FloatType x = (1. - ti) * x0 + ti * x1;
+      points.push_back(axom::primal::Point<FloatType, 2>({x, y}));
+    }
+  }
```
[the remark this answers is not in the document]
Comment (BradWhitlock) [reply src/axom/spin/tests/spin_bvh.cpp:1336]: I could use linspace.

File: src/axom/primal/geometry/BoundingBox.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kennyweiss) [reply src/axom/primal/geometry/BoundingBox.hpp:511]: Thanks -- in that case, this makes a lot more sense to me.

I agree that this seems like our current best way to copy the data when the types are potentially different, and that it is also pretty awkward / nonideal.

[the remark this answers is not in the document]
Comment (BradWhitlock) [reply src/axom/primal/geometry/BoundingBox.hpp:511]: No argument here ;)

File: src/axom/primal/geometry/BoundingBox.hpp [hunk 9]
```diff
@@ -529,14 +529,16 @@ AXOM_HOST_DEVICE BoundingBox<T, NDIMS>& BoundingBox<T, NDIMS>::expand(T expansio
 template <typename T, int NDIMS>
 AXOM_HOST_DEVICE BoundingBox<T, NDIMS>& BoundingBox<T, NDIMS>::scale(double scaleFactor)
 {
-  const PointType midpoint = getCentroid();
-  const VectorType r = static_cast<T>(scaleFactor * 0.5) * range();
-
-  m_min = PointType(midpoint.array() - r.array());
-  m_max = PointType(midpoint.array() + r.array());
+  if(this->isValid())
```
[the remark this answers is not in the document]
Comment (BradWhitlock) [reply src/axom/primal/geometry/BoundingBox.hpp:555]: I added one.

File: src/axom/spin/tests/spin_bvh.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (BradWhitlock) [reply src/axom/spin/tests/spin_bvh.cpp]: I changed the test so it uses linspace.

File: src/axom/spin/tests/spin_bvh.cpp (hunk 2 above)
Comment (BradWhitlock) [reply src/axom/spin/tests/spin_bvh.cpp]: Done.

File: src/axom/spin/tests/spin_bvh.cpp (hunk 3 above)
Comment (BradWhitlock) [reply src/axom/spin/tests/spin_bvh.cpp]: I made a symbol that is more descriptive.

File: src/axom/spin/policy/LinearBVH.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (kennyweiss) [reply src/axom/spin/policy/LinearBVH.hpp:63]: Thanks for checking. That seems reasonable, but we should probably add a note in the doxygen that `centroids` really only make sense for valid/non-empty bounding boxes.

File: src/axom/spin/tests/spin_bvh.cpp [hunk 10]
```diff
@@ -1306,6 +1306,162 @@ void check_find_points_zip2d()
...
+    /// Squared distance to query point
+    double minSqDist {numerics::floating_point_limits<double>::max()};
+    /// Index within mesh of closest element
+    int minElem {INVALID_ELEMENT_INDEX};
+  };
+
+  // Initialize the BVH with the points.
+  axom::IndexType npts = none ? 0 : points.size();
+  BoxType* bboxes =
+    axom::allocate<BoxType>(npts ? npts : 1,  // do not allocate 0 elements
```
Comment (kennyweiss) [inline src/axom/spin/tests/spin_bvh.cpp:1378]: Out of curiosity, did you get an error for `axom::allocate` with size `0`?
It's supposed to work (I think).

Comment (kennyweiss) [reply src/axom/spin/tests/spin_bvh.cpp:1378]: I didn't see explicit tests for it, but I did for `axom::reallocate(0)`, which is implemented in terms of `axom::allocate(0)`:

https://github.com/LLNL/axom/blob/1393b47f9c6110776ff1e2c4b86e5f2b62f1ed35/src/axom/core/memory_management.hpp#L221-L252

Comment (BradWhitlock) [reply src/axom/spin/tests/spin_bvh.cpp:1378]: I thought I coded around trying to allocate 0 elements, which is something I've always tried to avoid, probably from my C days.

Comment (kennyweiss) [reply src/axom/spin/tests/spin_bvh.cpp:1378]: We worked though the `allocate(0)`/`reallocate(0)` since it came up in practice with blueprint meshes. 
E.g. in a distributed mesh, some ranks have 0 boundary elements.

File: src/axom/spin/internal/linear_bvh/bvh_traverse.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (publixsubfan) [reply src/axom/spin/internal/linear_bvh/bvh_traverse.hpp:97]: I think we could move checks like these into the predicate functions defined in BVH.hpp:

https://github.com/LLNL/axom/blob/47a73d80f35b8c1574c8583d31c5e3fd1ee184d5/src/axom/spin/BVH.hpp#L519-L523

That being said, is there a case where we are returning `true` for a check involving an invalid box? I thought we had checks in the Spin intersection primitives for bounding box validity.

File: src/axom/primal/tests/primal_boundingbox.cpp [hunk 11]
```diff
@@ -491,6 +503,13 @@ TEST(primal_boundingBox, bb_scale)
   QBBox bbox4(bbox);
   bbox4.scale(-1.);
   EXPECT_EQ(bbox, bbox4);
+
+  // Show that scaling an invalid bounding box has no effect.
+  QBBox bbox5, bbox6;
+  EXPECT_FALSE(bbox5.isValid());
+  EXPECT_FALSE(bbox6.isValid());
+  bbox5.scale(1.5);
+  EXPECT_EQ(bbox5, bbox6);
```
Comment (gunney1) [inline src/axom/primal/tests/primal_boundingbox.cpp:512]: Is this a new thing, that invalid boxes compare equal?  Just asking because it could make sense that they shouldn't compare equal.  Maybe we have a precedence for this already, but I didn't see this behavior mentioned in the doxygen comments.

File: src/axom/spin/internal/linear_bvh/bvh_traverse.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (gunney1) [reply src/axom/spin/internal/linear_bvh/bvh_traverse.hpp:97]: @publixsubfan , I was wondering the same thing.

File: src/axom/spin/policy/LinearBVH.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (gunney1) [reply src/axom/spin/policy/LinearBVH.hpp:63]: I don't think `getCentroid` returns zero for an invalid box.  It returns some form of infinity.  `squared_distance` seems to treat it as (0,0) though.  This could be confusing for users.

File: src/axom/spin/tests/spin_bvh.cpp [hunk 12]
```diff
@@ -1306,6 +1306,162 @@ void check_find_points_zip2d()
...
+  return points;
+}
+
+//---------------------------------------------------------------------------
+
+/*!
+ * \brief Tests inserting a single point into the BVH and querying a set of
+ *        points against it, using a pattern seen in DistributedClosestPoint.
+ *        The test demonstrates current_node that the BVH returns an in range
+ *        value to the checkMinDist lambda.
```
Comment (gunney1) [inline src/axom/spin/tests/spin_bvh.cpp:1343]: Something's grammatically off about this statement.

**Review by gunney1 — APPROVED** [review]
Thank you @BradWhitlock

File: src/axom/spin/tests/spin_bvh.cpp [hunk 13]
```diff
@@ -1306,6 +1306,162 @@ void check_find_points_zip2d()
...
+    /// Index within mesh of closest element
+    int minElem {INVALID_ELEMENT_INDEX};
+  };
+
+  // Initialize the BVH with the points.
+  axom::IndexType npts = none ? 0 : points.size();
+  BoxType* bboxes =
+    axom::allocate<BoxType>(npts ? npts : 1,  // do not allocate 0 elements
+                            axom::execution_space<ExecSpace>::allocatorID());
+  for(axom::IndexType i = 0; i < npts; i++) bboxes[i] = BoxType(points[i]);
```
Comment (white238) [inline src/axom/spin/tests/spin_bvh.cpp:1380]: Please use braces on single line if/for's.

https://axom.readthedocs.io/en/develop/docs/sphinx/coding_guide/sec09_format.html#conditional-statements-and-loops

**Review by rhornung67 — APPROVED** [review]
